models/Device.js

```diff
@@ -30,25 +30,110 @@ module.exports = function(sequelize, DataTypes) {
     },
   };
 
-  var allForUser = async function(user) {
-    var deviceIds = await user.getDeviceIds();
-    var userGroupIds = await user.getGroupsIds();
-    return this.findAndCount({
-      where: { "$or": [
-        {GroupId: {"$in": userGroupIds}},
-        {id: {"$in": deviceIds}},
-      ]},
+  const models = sequelize.models;
+
+  /**
+  * Adds/update a user to a Device, if the given user has permissino to do so.
+  * The authenticated user must either be admin of the group that the device
+  * belongs to, an admin of that device, or a superuser.
+  */
+  const addUserToDevice = async function(authUser, deviceId, userToAddId, admin) {
+    const device = await models.Device.findById(deviceId);
+    const userToAdd = await models.User.findById(userToAddId);
+    if (device == null || userToAdd == null) {
+      return false;
+    }
+    // Return false if the user doesn't have permission.
+    const isGroupAdmin = await models.GroupUsers.isAdmin(device.groupId, authUser.id);
+    const isDeviceAdmin = await models.DeviceUsers.isAdmin(device.id, authUser.id);
+    if (!authUser.superuser && !isGroupAdmin && !isDeviceAdmin) {
+      return false;
+    }
+
+    // Get association if already there and update it.
+    var deviceUser = await models.DeviceUsers.findOne({
+      where: {
+        DeviceId: deviceId,
+        UserId: userToAdd.id,
+      }
+    });
```

Comment:
Instead of doing this direct query, I wonder if you could access the deviceUsers association via a property on the `device` that's already been loaded at the top of the function. sequelize adds properties to model objects for each association.

Reply:
I think so. If you say to include Users in the device query.

Reply:
Ok. Up to you whether you leave it as is or go via the association.